### PR TITLE
feat: invalidate older OTPs

### LINKED ISSUE #15: Expire existing OTP if a newer one is generated
**Describe the bug**
Currently, if we generate the OTP while logging in twice, we will get two OTPs which is expected. What is not expected is the older OTP to still work.

**To Reproduce**
1. Hit the `send-otp` endpoint with your email twice
2. Use the two OTPs one by one and see that both are working

**Expected behavior**
The older OTP shouldn't work

### PR DESCRIPTION
**Pull request by sharmaudit124**
## Type
bug_fix

___

## Description
This PR primarily addresses a bug where multiple OTPs could be valid at the same time for a single user. The changes made ensure that only the most recent OTP remains valid. The key changes are:
- A new private method `invalidateOldOtps` has been added to the `PrismaRepository` class. This method finds and deletes any old, unexpired OTPs associated with a user.
- The `createOtp` method in the `PrismaRepository` class now calls `invalidateOldOtps` before creating a new OTP, ensuring that only the most recent OTP is valid.
- Several files have been reformatted for better readability.

___

## PR changes walkthrough
<table><thead><tr><th></th><th>Relevant files&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </th></tr></thead><tbody><tr><td><strong>Formatting</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>admin.guard.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/auth/admin.guard.ts<br><br>

**The function signature for `canActivate` has been <br>reformatted for better readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-f48320353e9b903421b045677cdfd3c25111351f43dcea7b62fdd5dff587ff39"> +3/-1</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>auth.controller.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/auth/auth.controller.ts<br><br>

**The function signature for `validateOtp` has been <br>reformatted for better readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-e7e18e8f707440ebab6679949c480961dd946e5eea5f39f8104546da07f9d301"> +4/-1</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>auth.guard.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/auth/auth.guard.ts<br><br>

**The import statements have been reformatted for better <br>readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-db4ffd7a08656626f522fe8b4383e2ecf3baada3495914ab1af05bca26acd5f7"> +6/-1</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>auth.service.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/auth/auth.service.ts<br><br>

**The import statements and function signatures for `sendOtp` <br>and `validateOtp` have been reformatted for better <br>readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-7e9ac17d7e6417f73125395406cff59ad6eee28fab0bcca2993d72e2e439fae6"> +25/-5</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>user.decorator.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/decorators/user.decorator.ts<br><br>

**The function signature for `CurrentUser` has been <br>reformatted for better readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-476278995520cb19ae4b3abcb8f27cf5f4d02d6be56b3281b17dcfcbbe2d53b8"> +8/-6</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>main.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/main.ts<br><br>

**The `info`, `error`, `warn` methods and `logger.log` in the <br>`bootstrap` function have been reformatted for better <br>readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-32c5294dd7ca131397ef57a87c3e0cf38ea5c2e4ca0061c602e73f7f00514c1a"> +18/-4</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mail.resend.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/resend/services/mail.resend.ts<br><br>

**The function signature for `sendEmail` has been reformatted <br>for better readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-ce34d66b2dee167e082fd748f86244ec90eb9a708474672d7e33dbd3c6329a2f"> +5/-1</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>update.user.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/user/dto/update.user/update.user.ts<br><br>

**The file has been reformatted for better readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-2b50181fa0b5500dc29e7db78784db89bbf299cf42ccbf098db0c18cf1073875"> +4/-4</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Bug_fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>prisma.repository.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/prisma/prisma.repository.ts<br><br>

**A new private method `invalidateOldOtps` has been added to <br>invalidate old OTPs for a user. 
The `createOtp` method now <br>calls `invalidateOldOtps` before creating a new OTP.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-54a64b4b32a258b1a069196279309a21afc354a47579a3e14d51df18b471d461"> +46/-0</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>update.user.spec.ts&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/src/user/dto/update.user/update.user.spec.ts<br><br>

**The test file has been reformatted for better readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-25fdb1858305450e2da3428c02e060f9fef85d5852fdb3dce03f88fc6fff9a8c"> +4/-4</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>tsconfig.spec.json&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        apps/api/tsconfig.spec.json<br><br>

**The `include` array has been reformatted for better <br>readability.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/37/files#diff-21b3d5855a77c7d021343eb655f3922fc01db202d69a252face8733684a2ce29"> +6/-1</a></td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

## Description

Fixes the bug described in the issue where, upon hitting the send-otp endpoint multiple times with the same email, two or more OTPs were generated, and the older OTPs were  functional. 
Implemented a solution that deletes the older OTPs from the database. This ensures that only the most recent OTP remains valid for authentication.

Changes made in prisma.repository.ts file.
<br />

Fixes #15 

## Dependencies

_NONE_

## Future Improvements

_NONE_

## Mentions

@rajdip-b  please review this.

## Screenshots of relevant screens

_NONE_

## Developer's checklist

- [X] My PR follows the style guidelines of this project
- [X] I have performed a self-check on my work

**If changes are made in the code:**

- [X] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [X] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [X] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

### CODE REVIEW

File: apps/api/src/prisma/prisma.repository.ts [hunk 1]
```diff
@@ -142,6 +142,7 @@ export class PrismaRepository {
     expiresAfter: number
   ): Promise<Otp> {
     const timeNow = new Date()
+    await this.invalidateOldOtps(email)
```
Comment (rajdip-b) [inline apps/api/src/prisma/prisma.repository.ts:145]: Please move this to a cron job. Create a cron function in the auth service itself. Ref: https://docs.nestjs.com/techniques/task-scheduling. I would like the old OTPs to be deleted every hour. I believe there is a thing called `CronInterval.HOUR` or something similar in the reference I shared.

File: apps/api/src/prisma/prisma.repository.ts [hunk 2]
```diff
@@ -155,6 +156,51 @@ export class PrismaRepository {
...
+   * ```
+   */
+  private async invalidateOldOtps(email: string): Promise<void> {
+    const user = await this.prisma.user.findUnique({
+      where: {
+        email
+      }
+    })
+    if (user) {
+      const oldOtpCodesToDelete = await this.prisma.otp
```
Comment (rajdip-b) [inline apps/api/src/prisma/prisma.repository.ts:179]: I think you can spare yourself (and us!) some time complexity, and pass the conditions directly in `prisma.otp.deleteMany` call.

**Review by rajdip-b — CHANGES_REQUESTED** [review]
Nice work overall @sharmaudit124 . I have added my comments. Please do go through them and make the necessary changes. Also, it would be really nice if you can mention in the issue that you would like to take it up! It always helps us a lot in managing the progress!

**Comment by rajdip-b** [comment]
Hey @sharmaudit124 ! sorry i got confused with the feature! What you did was totally perfect. the thing I mentioned is related to some other issue! i will make the review again and merge it

**Comment by rajdip-b** [comment]
Okei I have updated what needs to be resolved. Just make sure you take a pull from the main branch before you start working again!

**Comment by sharmaudit124** [comment]
> Okei I have updated what needs to be resolved. Just make sure you take a pull from the main branch before you start working again!

Okay, Thank you! @rajdip-b  I will definitely mention in the issue that I would like to take that! and will take pull request from main branch.

**Comment by sharmaudit124** [comment]
> Hey @sharmaudit124 ! Sorry, I got confused with the feature! What you did was totally perfect. The thing I mentioned is related to some other issue! I will make the review again and merge it.
> 
> **EDIT:** You can refer to #42 for the thing I mentioned. I will be more than happy to assign that to you as well!

Hey @rajdip-b No worries at all! I appreciate your clarification. It happens to the best of us. Thanks for pointing out the related issue #42, and I'm more than happy to take that on as well. Let me know if there's anything specific you'd like me to focus on in that task.

**Comment by rajdip-b** [comment]
That's much appreciated! I have assigned you that issue, do get on with it once you are finished with this one,

**Comment by rajdip-b** [comment]
P.S., do lemme know when you are done with this one!

**Comment by sharmaudit124** [comment]
> P.S., do lemme know when you are done with this one!

Sure! I will let you know soon!

File: apps/api/src/prisma/prisma.repository.ts (hunk 2 above)
Comment (sharmaudit124) [reply apps/api/src/prisma/prisma.repository.ts]: I have updated the changes, please review it @rajdip-b

**Comment by rajdip-b** [comment]
Hey @sharmaudit124 , all seems good. I'll merge it soon.

**Review by rajdip-b — APPROVED** [review]
Hey bro! This looks perfect now! I am merging you code. Two thing's I would like to state:
- Would you be interested to work on #42 ?
- When you are developing some feature, don't do it on your main branch. Instead, create a new branch from main, do the dev there, and make a PR from that branch to our main!

**Comment by sharmaudit124** [comment]
> Hey bro! This looks perfect now! I am merging you code. Two thing's I would like to state:
> 
> * Would you be interested to work on [Create scheduler to delete expired OTPs #42](https://github.com/keyshade-xyz/keyshade/issues/42) ?
> * When you are developing some feature, don't do it on your main branch. Instead, create a new branch from main, do the dev there, and make a PR from that branch to our main!

Cool, got it, moving forward will create a separate branch of my forked repo too and will then raise pr accordingly!
The issue #42 is currently assigned to someone else. In the meantime, I'll explore other potential tasks.

**Comment by rajdip-b** [comment]
I have merged it for now, but do keep that branch thing in your mind from the next  PRs!

**Pull request merged**